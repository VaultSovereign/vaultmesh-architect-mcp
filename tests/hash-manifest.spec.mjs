import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import { fileURLToPath } from 'node:url';
import { tmpdir } from 'node:os';
import { mkdtemp, readFile } from 'node:fs/promises';
import { join } from 'node:path';
import { execa } from 'execa';
import { makeFixture } from './manifest.fixture.mjs';

async function rpc(server, method, params = {}) {
  const payload = JSON.stringify({ jsonrpc: '2.0', id: Date.now(), method, params }) + '\n';
  server.stdin.write(payload);
  const { value } = await new Promise((resolve) => {
    server.stdout.once('data', (buf) => {
      resolve({ value: JSON.parse(String(buf)) });
    });
  });
  if (value.error) throw new Error(value.error.message || 'RPC error');
  return value.result;
}

describe('compute_merkle_root â†’ golden ordering', () => {
  let work, server;

  beforeAll(async () => {
    work = await mkdtemp(join(tmpdir(), 'vm-'));
    await makeFixture(work);
    server = execa('node', ['server.js', '--stdio'], {
      cwd: fileURLToPath(new URL('../', import.meta.url)),
      env: { ...process.env, VM_WORKDIR: work, DRY_RUN: 'true', MCP_HEARTBEAT: '0' },
      stdin: 'pipe', stdout: 'pipe', stderr: 'pipe'
    });
    // Initialize per MCP spec
    await rpc(server, 'initialize', {
      protocolVersion: '2025-06-18',
      capabilities: {},
      clientInfo: { name: 'vaultmesh-tests', version: '0.0.0' }
    });
  }, 30000);

  afterAll(async () => {
    if (server) {
      try {
        server.stdin.end();
        await Promise.race([
          server.catch(() => {}),
          new Promise((resolve) => setTimeout(resolve, 500))
        ]);
        if (server.exitCode == null) {
          server.kill('SIGTERM');
        }
      } catch {}
    }
  });

  it('emits stable, sorted manifest and root', async () => {
    const outPath = 'manifests/hash-manifest.json';
    await rpc(server, 'tools/call', {
      name: 'compute_merkle_root',
      arguments: { root: '.', out: outPath }
    });
    const manifest = JSON.parse(await readFile(join(work, outPath), 'utf8'));
    // Snapshot only ordering-relevant parts
    const view = {
      merkleRoot: manifest.merkleRoot,
      files: manifest.files.map((e) => ({ path: e.path, hash: e.hash }))
    };
    expect(view).toMatchSnapshot();
  });
});
