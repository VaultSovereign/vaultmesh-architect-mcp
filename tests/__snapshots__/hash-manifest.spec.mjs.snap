// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compute_merkle_root â†’ golden ordering > emits stable, sorted manifest and root 1`] = `
{
  "files": [
    {
      "hash": "5ddbce254c08372e429a250112c6f4593868687ab01e9a126193e5a83560362b",
      "path": ".dotfile",
    },
    {
      "hash": "b6a98d9ce9a2d9149288fa3df42d377c3e42737afdcdaf714e33c0a100b51060",
      "path": "a/alpha.txt",
    },
    {
      "hash": "2088d0c4b41022d90f663fa8d8156cb525241b55d30ecdf922c38f94f7efda4c",
      "path": "a/zeta.txt",
    },
    {
      "hash": "f2c82decdd7181cf98945929a62598db7e6b477e11f6e0eb0ae97020eff151ad",
      "path": "b/beta.txt",
    },
    {
      "hash": "054edec1d0211f624fed0cbca9d4f9400b0e491c43742af2c5b0abebf0c990d8",
      "path": "b/img.bin",
    },
  ],
  "merkleRoot": "8a99cc7c12ffb388ef82e1215d6e7eb1cc667b3c1c1c406feed9f236fcd11fe1",
}
`;
